### DISCUSSION TITLE
Octocrab and 204 responses

### DISCUSSION THREAD

**Discussion by manchicken** [XAMPPRocky/octocrab#506] [Ideas]
There are a number of endpoints which return a 204, and it kinda looks like a bunch of them are currently broken. The one test I added in #503 as part of issue #504 shows how the standard `FromResponse::from_response()` approach doesn't account for this.

There appear to be 263 endpoints which could return a 204, so while I am tempted to just fix the ones I'm trying to use, it kinda seems like we need a more comprehensive approach.

Here are the endpoints I found a 204 on, per [GitHub's OpenAPI spec](https://raw.githubusercontent.com/github/rest-api-description/main/descriptions/ghec/ghec.yaml): [findings.csv](https://github.com/XAMPPRocky/octocrab/files/13727504/findings.csv)

**Comment by manchicken** [comment]
I started tracking all of these here: https://github.com/users/manchicken/projects/4/views/1

I did a quick write-up of my process here: https://dev.to/manchicken/tackling-a-large-volume-of-simple-problems-in-an-open-source-project-5fn3

**Comment by manchicken** [comment]
I've gone through a bunch of the API endpoints with 204s, and the good news is that so far I haven't found any more that are broken. The bad news is that there are ~164 left to sift through. The biggest thing I'm finding is that there are a lot of gaps in API coverage.

@XAMPPRocky, do you have any thoughts or documentation regarding your philosophy on which API endpoints get a semantic function and which ones don't? If—after I finish this little 204 audit—I wanted to add some, is there any guidance on which ones would be wanted and which wouldn't?

Many thanks for the opportunity to help.

**Reply by XAMPPRocky** [reply]
All endpoints get a semantic function, the only reason one doesn't is that no one has added it yet.

**Reply by manchicken** [reply]
Thanks for the quick answer! After I'm done with the 204 audit I'll find ones to add.

**Comment by manchicken** [comment]
@dmgorsky, I see a bunch of the issues you made to add semantic API functions for uncovered endpoints. I'm working on adding tests for these 204 cases (and fixes where needed).

Would you like to collaborate on putting prioritizing the endpoints that need to be added? If our maintainer says all endpoints get a semantic function then this could be a fun series of little projects.

**Reply by dmgorsky** [reply]
TBH I just started with whatever :D i.e. https://github.com/XAMPPRocky/octocrab/issues/529 (checked in issue what's completed here https://github.com/dmgorsky/octocrab/tree/feature/checks)
I am ready to collaborate on prioritizing! (went inspecting your links)

**Reply by manchicken** [reply]
Marvelous! I'm about to wrap up this one: https://github.com/users/manchicken/projects/4/views/1?pane=issue&itemId=49890596

Do you think it would be helpful to set up a project like I did for the 204s? I wonder if that would make it easier for us to help folks pick them up and move forward.

**Comment by manchicken** [comment]
With that last PR, I’ve covered all of the 204 cases for which semantic functions exist. 🥳

**Reply by manchicken** [reply]
Test coverage before this project (calculated using `cargo tarpaulin`):

```
22.94% coverage, 484/2110 lines covered
```

Test coverage after this project (calculated using `cargo tarpaulin`):

```
37.63% coverage, 1433/3808 lines covered, +14.69% change in coverage
```

**Comment by manchicken** [comment]
@dmgorsky I finished the project I was working on. Do you have a process you were following for adding functionality? I'm gonna see if I can find an endpoint among the ones you identified as missing to implement.

If you—or anybody else—has a prioritized list, I'd be open to addressing these in an order.
